Kconfig

```diff
@@ -2124,7 +2132,7 @@ source "openamp/Kconfig"
 endmenu
 
 menu "Application Configuration"
-source "$APPSDIR/Kconfig"
+source "$APPSBINDIR/Kconfig"
```

Comment:
Is this correct? 

Please add a comments to the declarations of APPSDIR, APPSBINDIR and BINDIR to give the context.

Please state what they are and what they are used for including lifecycle (source, created by, precious or deleted etc)

Reply:
NuttX's kconfig system is different from other RTOS. Some Kconfigs are dynamically generated through different configurations of the system. For example:

https://github.com/apache/nuttx/blob/1cb11968b7b207806cbfef9012943efae9819258/Kconfig#L2127

https://github.com/apache/nuttx/blob/1cb11968b7b207806cbfef9012943efae9819258/arch/Kconfig#L232


In the latest cmake environment, we will support out-of-tree build, which means that the "out" directory can be specified through compilation parameters, and multiple configurations can be built with same code base:

```bash
cmake -B nsh -DBOARD_CONFIG=sim/nsh -GNinja
cmake -B nsh2 -DBOARD_CONFIG=sim/nsh2 -GNinja
```

cmake can help us solve the source code out-of-tree support, but the kconfig system does not work, so we optimize the dynamically created Kconfig to a version with configurable paths, so that when compiling different configurations, only the "out" directory generate a temporary Kconfig,
To achieve the purpose of decoupling temporary files and nuttx directories:


```bash
nuttx/out/nsh/apps$ ls 
audioutils           crypto    graphics                                            _home_archer_code_nuttx_n8_apps_fsutils_Kconfig       _home_archer_code_nuttx_n8_apps_math_Kconfig       include       mlearning  testing
boot                 dummy.c   _home_archer_code_nuttx_n8_apps_audioutils_Kconfig  _home_archer_code_nuttx_n8_apps_games_Kconfig         _home_archer_code_nuttx_n8_apps_mlearning_Kconfig  industry      modbus     wireless
builtin              examples  _home_archer_code_nuttx_n8_apps_boot_Kconfig        _home_archer_code_nuttx_n8_apps_gpsutils_Kconfig      _home_archer_code_nuttx_n8_apps_netutils_Kconfig   interpreters  netutils
canutils             fsutils   _home_archer_code_nuttx_n8_apps_canutils_Kconfig    _home_archer_code_nuttx_n8_apps_graphics_Kconfig      _home_archer_code_nuttx_n8_apps_system_Kconfig     Kconfig       nshlib
CMakeFiles           games     _home_archer_code_nuttx_n8_apps_crypto_Kconfig      _home_archer_code_nuttx_n8_apps_industry_Kconfig      _home_archer_code_nuttx_n8_apps_testing_Kconfig    libapps.a     platform
cmake_install.cmake  gpsutils  _home_archer_code_nuttx_n8_apps_examples_Kconfig    _home_archer_code_nuttx_n8_apps_interpreters_Kconfig  _home_archer_code_nuttx_n8_apps_wireless_Kconfig   math          system
```

Of course, if you are not familiar with kconfig, you can try PR https://github.com/apache/nuttx/pull/6718 to experience it.

In addition, this change will not break ./tools/configure.sh, you can see that ci can pass normally, so if you don't use cmake, you don't need to care about this PR.

Reply:
@davids5 since apps/Kconfig is generated dynamically. Changing $APPSDIR to $APPSBINDIR could move the generated Kconfig to the different directory, so it's the required step to support the out of tree build. BTW, since APPSBINDIR is same as APPSDIR by default. The behaviour keeps the same as before, but APPSBINDIR will set to a different path in cmake build system for out of tree buid.

Reply:
@davids5 do you have more concern?

Reply:
@xiaoxiang781216 Yes. I asked for comments in the Kconfig file, they were not added